Block Themes: Add Infrastructure

ockham opened the pull request:
Work in progress. Attempts to add the required infrastructure to render block-based themes to Core.

Trac ticket: https://core.trac.wordpress.org/ticket/54335

## TODO

- [x] Add Template Part CPT and TP area taxonomy registration, unless we decide that it's not strictly needed to render block themes, and defer to a separate PR. Note that I've reverted for now (437a16a) since it caused errors in wp-admin.
- [x] Add minified CSS to Template Part block.
- [x] Add unit test coverage (carry over from Gutenberg).
- [x] Remove `gutenberg_` prefixes and `_gutenberg_` infixes.
- [x] Add/update PHPDocs to cover added/changed functionality; add `@since` items where needed.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

youknowriad commented on src/wp-includes/block-template-part-utils.php:
```diff
@@ -0,0 +1,180 @@
+/**
+ * Fixes the label of the 'wp_template_part' admin menu entry.
+ */
+function gutenberg_fix_template_part_admin_menu_entry() {
```
I'm not sure the admin menu is one we want to include, I'll remove this for now.

ockham replied:
Yeah, I wasn't too sure about that myself. Makes sense to apply the same logic as we're doing with templates.

youknowriad commented on src/wp-includes/blocks/template-part.php:
```diff
@@ -0,0 +1,155 @@
+<?php
```
Ideally this file shouldn't be touched manually on Core. That said, I had to modify calls to functions referring `gutenberg_` variations (I also renamed some functions to clarify their purpose). We should find a way to bring these changes back to Gutenberg and make sure the automatic renaming works properly.

ockham replied:
TBH I wasn't quite sure what the process is to bring a new block from GB into Core. Do we have docs on this? (Also see my remaining TODO, "Add minified CSS to Template Part block.")

youknowriad replied:
These come automatically when you do `npm install`. I think the first time you add a dynamic block though, you might have to whitelist it somewhere though.

ockham replied:
I thought I did that here 🤔 
```suggestion
<?phphttps://github.com/WordPress/wordpress-develop/pull/1796/files#diff-777759e98c224019a699e8294c23bccc34453c91b873a12cb78a828487cd5f15R56
```

> Ideally this file shouldn't be touched manually on Core. That said, I had to modify calls to functions referring gutenberg_ variations (I also renamed some functions to clarify their purpose). We should find a way to bring these changes back to Gutenberg and make sure the automatic renaming works properly.

How would we do this? Right now, the build step prefixes a few functions in this file with `gutenberg_`, which causes both the frontend and backend to fail 🤔  It'd be great if we could somehow unblock this...

cc/ @ntsekouras in case you wanna look into this 😄

youknowriad replied:
We need to update the code in the Gutenberg plugin to avoid relying on prefixed functions. The build step prefixes for the plugin but the file stay as (no prefix) for Core. It will require a package release though to make it back to here.

ntsekouras replied:
>We need to update the code in the Gutenberg plugin to avoid relying on prefixed functions

I can try that. I'll share a PR to verify if my approach is correct.

youknowriad commented:
I've merged template-part utils and template utils because the template-utils already worked for both. We just missed some functions.

youknowriad commented:
>  Add minified CSS to Template Part block.

What is this item about @ockham I think there's no stylesheet for this block if I'm not wrong?

youknowriad commented:
Ok, I think this one is actually ready, the only issue is the manual modifications to the template part blocks which is breaking some tests (unit, e2e tests, npm).

ntsekouras commented on src/wp-includes/block-template-utils.php:
```diff
@@ -53,13 +525,40 @@ function _build_template_result_from_post( $post ) {
+	 *
+	 * Return a non-null value to bypass the WordPress queries.
+	 *
+	 * @since 10.8
```
I guess this should be `5.9`, no?

ntsekouras commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1,23 +1,481 @@
+ *
+ * @return array Template.
+ */
+function _get_block_templates_files( $template_type ) {
```
Should this be `_get_block_template_files`? (no plural in `templates`)

youknowriad replied:
I hesitated but I choice "templates" because it's multiple templates while `template-files` could mean mutiple files for a single template. Not entirely confident with my choice, that said, it's not very important because it's a private function.

ntsekouras commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -395,6 +407,16 @@ protected function prepare_item_for_database( $request ) {
 
+		if ( 'wp_template_part' === $this->post_type ) {
+			if ( isset( $request['area'] ) ) {
+				$changes->tax_input['wp_template_part_area'] = filter_template_part_area( $request['area'] );
```
I guess this and below call should be `_filter_block_template_part_area`, no?

youknowriad replied:
Good catch, I updated them.

ntsekouras commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -27,27 +49,170 @@ public static function wpSetUpBeforeClass() {
+	 * Should retrieve the template from the theme files.
+	 */
+	function test_get_block_template_from_file() {
+		$this->markTestIncomplete();
```
Why skipped this? I'm having the same issue in the GB PR.

youknowriad replied:
Here It's skipped because there's no FSE theme we can active on Core yet.

ockham replied:
WP 5.9 will [include Twenty Twenty Two](https://core.trac.wordpress.org/ticket/54318) though 🤔  I guess we should file an issue or ticket somewhere to re-enable these tests once we have that theme available in Core.

noisysocks commented:
> Ok, I think this one is actually ready, the only issue is the manual modifications to the template part blocks which is breaking some tests (unit, e2e tests, npm).

It's a bit of a chicken and egg problem because https://github.com/WordPress/wordpress-develop/pull/1804 which updates the blocks properly depends on this PR 🙂

I guess we will need to just skip the tests or commit this with failing tests and then immediately follow up with commits for https://core.trac.wordpress.org/ticket/54336 and https://core.trac.wordpress.org/ticket/54337.

youknowriad commented:
@noisysocks yeah, I'm also fine if we include this patch in the packages update one, whatever works for you.

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1,23 +1,481 @@
+		),
+	);
+
+	/**
+	 * Filters the list of allowed template part area values.
+	 *
+	 * @param array $default_areas An array of supported area objects.
+	 */
```
Filters also need an `@since`.
```suggestion
	/**
	 * Filters the list of allowed template part area values.
	 *
	 * @since 5.9.0
	 *
	 * @param array $default_areas An array of supported area objects.
	 */
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1,23 +1,481 @@
+		),
+	);
+
+	/**
+	 * Filters the list of template types.
+	 *
+	 * @param array $default_template_types An array of template types, formatted as [ slug => [ title, description ] ].
+	 *
+	 * @since 5.x.x
+	 */
```
```suggestion
	/**
	 * Filters the list of template types.
	 *
	 * @since 5.9.0
	 *
	 * @param array $default_template_types An array of template types, formatted as [ slug => [ title, description ] ].
	 */
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1,23 +1,481 @@
+ */
+function _filter_block_template_part_area( $type ) {
+	$allowed_areas = array_map(
+		function ( $item ) {
```
For performance, closures now use `static` when not within an object that reference itself via `$this`. See changeset https://core.trac.wordpress.org/changeset/51657.

```suggestion
		static function ( $item ) {
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1,23 +1,481 @@
+		return $type;
+	}
+
+	/* translators: %1$s: Template area type, %2$s: the uncategorized template area value. */
+	$warning_message = sprintf( __( '"%1$s" is not a supported wp_template_part area value and has been added as "%2$s".' ), $type, WP_TEMPLATE_PART_AREA_UNCATEGORIZED );
```
During the 5.9.0 cycle, we've been moving the translators comment directly above it's function and putting longer lines of code like this on multiple lines (for improved readability):

```suggestion
	$warning_message = sprintf(
		/* translators: %1$s: Template area type, %2$s: the uncategorized template area value. */
		__( '"%1$s" is not a supported wp_template_part area value and has been added as "%2$s".' ),
		$type,
		WP_TEMPLATE_PART_AREA_UNCATEGORIZED
	);
```

Also `wp_template_part` in its current form could confuse translators. Consider a non-code version or adding another placeholder.

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1,23 +1,481 @@
+		get_template()   => get_template_directory(),
+	);
+	foreach ( $themes as $theme_slug => $theme_dir ) {
+		$file_path = $theme_dir . '/' . $template_base_paths[ $template_type ] . '/' . $slug . '.html';
```
What happens if the `$template_type` is not in the `$template_base_paths`? Will the value passed to it _always_ be `wp_template` or `wp_template_part`? If yes, then it's fine. Else, consider a guard clause at the start of the function such as:

```php
if ( 'wp_template' !== $template_type || 'wp_template_part' !== $template_type ) {
	return null;
}
```

noisysocks replied:
Will defer to @ockham here 🙂

ockham replied:
I think bailing early makes sense, since we don't totally control what template type is being passed here.

Fixed in f4d281c. (I changed the `||` to an `&&` though, since I think that's the right condition here.)

ockham replied:
GB counterpart PR: https://github.com/WordPress/gutenberg/pull/36318

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1,23 +1,481 @@
+
+	$template_files = array();
+	foreach ( $themes as $theme_slug => $theme_dir ) {
+		$theme_template_files = _get_block_templates_paths( $theme_dir . '/' . $template_base_paths[ $template_type ] );
```
Same question here as https://github.com/WordPress/wordpress-develop/pull/1796/files#r744135286.

ockham replied:
Fixed in f4d281c, see https://github.com/WordPress/wordpress-develop/pull/1796/files#r744951768.

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -88,26 +596,96 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
  *
- * @param string $id            Template unique identifier (example: theme_slug//template_slug).
- * @param string $template_type Optional. The template type (post type). Default 'wp_template'.
+ * @param string $id Template unique identifier (example: theme_slug//template_slug).
+ * @param array  $template_type wp_template or wp_template_part.
+ *
```
```suggestion
 * @param string $id                      Template unique identifier (example: theme_slug//template_slug).
 * @param array  $template_type Optional. Template type. Values are: `'wp_template'` or '`wp_template_part'`.
 *                                                       Default `'wp_template'`.
 *
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -131,12 +709,123 @@ function get_block_template( $id, $template_type = 'wp_template' ) {
+	 *
+	 * @since 5.9
+	 *
+	 * @param WP_Block_Template $block_template The found block template.
+	 * @param string $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array  $template_type wp_template or wp_template_part.
```
```suggestion
	 * @param WP_Block_Template $block_template     The found block template.
	 * @param string                         $id                             Template unique identifier (example: theme_slug//template_slug).
	 * @param array                          $template_type       Template type. Values are: `'wp_template'` or '`wp_template_part'`.
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -88,26 +596,96 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
+	 *
+	 * Return a non-null value to bypass the WordPress queries.
+	 *
+	 * @since 5.9
+	 *
+	 * @param WP_Block_Template|null $block_template Return block template object to short-circuit the default query,
+	 *                                               or null to allow WP to run it's normal queries.
+	 * @param string $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array  $template_type wp_template or wp_template_part.
```
```suggestion
	 * @since 5.9.0
	 *
	 * @param WP_Block_Template|null $block_template Return block template object to short-circuit the default query,
	 *                                               or null to allow WP to run its normal queries.
	 * @param string $id                             Template unique identifier (example: theme_slug//template_slug).
	 * @param array  $template_type                  Template type: `'wp_template'` or '`wp_template_part'`.
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -131,12 +709,123 @@ function get_block_template( $id, $template_type = 'wp_template' ) {
+ *
+ * @since 5.9.0
+ *
+ * @param string $id Template unique identifier (example: theme_slug//template_slug).
+ * @param array  $template_type wp_template or wp_template_part.
+ *
+ * @return WP_Block_Template|null File template.
```
```suggestion
  * @param string $id            Template unique identifier (example: theme_slug//template_slug).
 * @param array  $template_type Optional. Template type: `'wp_template'` or '`wp_template_part'`.
 *                              Default `'wp_template'`.
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -131,12 +709,123 @@ function get_block_template( $id, $template_type = 'wp_template' ) {
+	 *
+	 * Return a non-null value to bypass the WordPress queries.
+	 *
+	 * @since 5.9
+	 *
+	 * @param WP_Block_Template|null $block_template Return block template object to short-circuit the default query,
+	 *                                               or null to allow WP to run it's normal queries.
+	 * @param string $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array  $template_type wp_template or wp_template_part.
```
```suggestion
	 *
	 * @since 5.9.0
	 *
	 * @param WP_Block_Template|null $block_template Return block template object to short-circuit the default query,
	 *                                               or null to allow WP to run its normal queries.
	 * @param string $id                             Template unique identifier (example: theme_slug//template_slug).
	 * @param array  $template_type                  Template type: `'wp_template'` or '`wp_template_part'`.
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -131,12 +709,123 @@ function get_block_template( $id, $template_type = 'wp_template' ) {
+}
+
+/**
+ * Retrieves a single unified template object using its id.
+ * Retrieves the file template.
```
This confuses me. Should the second line be deleted?

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -131,12 +709,123 @@ function get_block_template( $id, $template_type = 'wp_template' ) {
+	/**
+	 * Filters the array of queried block templates array after they've been fetched.
+	 *
+	 * @since 5.9
+	 *
+	 * @param null|WP_Block_Template $block_template The found block template.
+	 * @param string $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array  $template_type wp_template or wp_template_part.
```
```suggestion
	 * @since 5.9.0
	 *
	 * @param WP_Block_Template $block_template The found block template.
	 * @param string            $id             Template unique identifier (example: theme_slug//template_slug).
	 * @param array             $template_type  Template type: `'wp_template'` or '`wp_template_part'`.
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -131,12 +709,123 @@ function get_block_template( $id, $template_type = 'wp_template' ) {
+ * @since 5.9.0
+ *
+ * @param string $part The template-part to print. Use "header" or "footer".
+ *
+ * @return void
```
`@return void` not necessary.
```suggestion
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -131,12 +709,123 @@ function get_block_template( $id, $template_type = 'wp_template' ) {
+ * Print the header template-part.
+ *
+ * @since 5.9.0
+ *
+ * @return void
```
```suggestion
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -131,12 +709,123 @@ function get_block_template( $id, $template_type = 'wp_template' ) {
+ * Print the footer template-part.
+ *
+ * @since 5.9.0
+ *
+ * @return void
```
```suggestion
```

hellofromtonya commented on src/wp-includes/blocks/template-part.php:
```diff
@@ -0,0 +1,155 @@
+ */
+
+/**
+ * Renders the `core/template-part` block on the server.
+ *
+ * @param array $attributes The block attributes.
+ *
+ * @return string The render.
+ */
```
```suggestion
 * Renders the `core/template-part` block on the server.
 *
 * @since 5.9.0
 * 
 * @param array $attributes The block attributes.
 * @return string The render.
 */
```

noisysocks replied:
This will need to be changed in Gutenberg as this file is copied from `@wordpress/block-library`.

ockham replied:
> This will need to be changed in Gutenberg as this file is copied from `@wordpress/block-library`.

Indeed. AFAICS, we don't use `@since` there (or at least not with Core versions), which makes sense (since it'd be misleading within the context of the GB repo and project, and its own versions).

hellofromtonya commented on src/wp-includes/blocks/template-part.php:
```diff
@@ -0,0 +1,155 @@
+			// render the corresponding file content.
+			$template_part_file_path = get_stylesheet_directory() . '/block-template-parts/' . $attributes['slug'] . '.html';
+			if ( 0 === validate_file( $attributes['slug'] ) && file_exists( $template_part_file_path ) ) {
+				$content = _inject_theme_attribute_in_block_template_content( file_get_contents( $template_part_file_path ) );
```
`file_get_contents()` returns `false` on failure. `_inject_theme_attribute_in_block_template_content()` expects a `string` data type. Within the deeper logic (i.e. in `WP_Block_Parser::next_token()) the content is passed as the 2nd argument to `preg_match`. That argument requires a non-nullable string.

To avoid possible PHP 8.1 deprecation notices and future PHP 9 fatal error, I'd suggest:

```suggestion
				$content = file_get_contents( $template_part_file_path );
				$content = is_string( $content ) && '' !== $content
					? _inject_theme_attribute_in_block_template_content(  $content )
					: '';
```

Also a micro-optimization to avoid further processing of an empty value.

hellofromtonya replied:
Note: May want to change the empty string value to `null` as the logic just below this code block bails out if there's no content.

noisysocks replied:
This will need to be changed in Gutenberg as this file is copied from `@wordpress/block-library`.

ockham replied:
> This will need to be changed in Gutenberg as this file is copied from `@wordpress/block-library`.

Yeah, this is sync'd from the `@wordpress/block-library` package upon build. For this reason, I won't apply the fix here but file a GB PR.

ockham replied:
https://github.com/WordPress/gutenberg/pull/36324

hellofromtonya commented on src/wp-includes/blocks/template-part.php:
```diff
@@ -0,0 +1,155 @@
+	return "<$html_tag $wrapper_attributes>" . str_replace( ']]>', ']]&gt;', $content ) . "</$html_tag>";
+}
+
+/**
+ * Returns an array of variation objects for the template part block.
+ *
+ * @return array Array containing the block variation objects.
+ */
```
```suggestion
/**
 * Returns an array of variation objects for the template part block.
 *
 * @since 5.9.0
 *
 * @return array Array containing the block variation objects.
 */
```

noisysocks replied:
This will need to be changed in Gutenberg as this file is copied from `@wordpress/block-library`.

ockham replied:
See my other comment about `@since` versions for GB-sync'd files. To confirm, I've grepped `since` in GB's `block-library` package directory, and I've only found one match. So I think we generally don't seem to add `@since` to dynamic blocks PHP in `block-library`.

hellofromtonya commented on src/wp-includes/blocks/template-part.php:
```diff
@@ -0,0 +1,155 @@
+	return $variations;
+}
+
+/**
+ * Registers the `core/template-part` block on the server.
+ */
```
```suggestion
/**
 * Registers the `core/template-part` block on the server.
 *
 * @since 5.9.0
 */
```

noisysocks replied:
This will need to be changed in Gutenberg as this file is copied from `@wordpress/block-library`.

ockham replied:
As stated elsewhere, it looks like dynamic blocks PHP shared from GB doesn't typically contain `@since` PHPDocs (as the version would be ambiguous, depending on whether it's used in a GB or Core context).

hellofromtonya commented on src/wp-includes/class-wp-block-template.php:
```diff
@@ -100,4 +100,12 @@ class WP_Block_Template {
+	/**
+	 * Whether a template is a custom template.
+	 *
+	 * @since 5.8.0
+	 * @var boolean
```
```suggestion
	 * @since 5.9.0
	 *
	 * @var boolean
```
nitpick can be changed at commit.

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -507,11 +529,19 @@ protected function get_available_actions() {
 	 */
 	public function get_collection_params() {
 		return array(
-			'context' => $this->get_context_param(),
```
Suggestion: Add an @since comment to the method's DocBlock to reflect the additions:

```
	 * @since 5.8.0
	 * @since 5.9.0 Added `'area'` and `'post_type'`.
```

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -596,6 +626,14 @@ public function get_item_schema() {
 			),
 		);
 
+		if ( 'wp_template_part' === $this->post_type ) {
```
Suggestion see https://github.com/WordPress/wordpress-develop/pull/1796/files#r744141163

hellofromtonya commented on src/wp-includes/taxonomy.php:
```diff
@@ -191,6 +191,25 @@ function create_initial_taxonomies() {
 			'show_in_rest'      => false,
 		)
 	);
+
```
Suggestion: Note this new taxonomy in the function's DocBlock like this:

```
 * @since 5.9.0 Added `'wp_template_part_area'` taxonomy.
```

Odd that other additions weren't already noted there. Note to myself to update that function's revisions DocBlock.

hellofromtonya commented on src/wp-includes/theme-templates.php:
```diff
@@ -1,5 +1,33 @@
 <?php
 
+/**
+ * Sets a custom slug when creating auto-draft template parts.
+ * This is only needed for auto-drafts created by the regular WP editor.
+ * If this page is to be removed, this won't be necessary.
+ *
+ * @param int $post_id Post ID.
```
```suggestion
 * Sets a custom slug when creating auto-draft template parts.
 *
 * This is only needed for auto-drafts created by the regular WP editor.
 * If this page is to be removed, this won't be necessary.
 *
 * @since 5.9.0
 *
```

hellofromtonya commented on src/wp-includes/theme-templates.php:
```diff
@@ -1,5 +1,33 @@
+	}
+
+	$terms = get_the_terms( $post_id, 'wp_theme' );
+	if ( ! $terms || ! count( $terms ) ) {
```
Fatal error in PHP 8.x
Warning in < PHP 8.0

How? if `get_the_terms()` returns `WP_Error` or `false`.

Why? The PHP native `count` requires a `Countable` or `array` type. See [PHP manual](https://www.php.net/manual/en/function.count.php) and [in action here](https://3v4l.org/c7nB7#v8.1rc3).

To guard against these problems, change the first conditional to check if what's returned is not an array:
```suggestion
	if ( ! is_array( $terms ) || ! count( $terms ) ) {
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -27,27 +49,170 @@ public static function wpSetUpBeforeClass() {
+			'wp_template'
+		);
+
+		$this->assertEquals( get_stylesheet() . '//single', $template->id );
+		$this->assertEquals( get_stylesheet(), $template->theme );
+		$this->assertEquals( 'single', $template->slug );
+		$this->assertEquals( 'publish', $template->status );
+		$this->assertEquals( 'theme', $template->source );
+		$this->assertEquals( 'Single Post', $template->title );
+		$this->assertEquals( 'Template used to display a single blog post.', $template->description );
+		$this->assertEquals( 'wp_template', $template->type );
```
With PHP getting more and more strict, an effort is ongoing to replace all `assertEquals` with `assertSame`.

```suggestion
		$this->assertSame( get_stylesheet() . '//single', $template->id );
		$this->assertSame( get_stylesheet(), $template->theme );
		$this->assertSame( 'single', $template->slug );
		$this->assertSame( 'publish', $template->status );
		$this->assertSame( 'theme', $template->source );
		$this->assertSame( 'Single Post', $template->title );
		$this->assertSame( 'Template used to display a single blog post.', $template->description );
		$this->assertSame( 'wp_template', $template->type );
```

In the event any of these assertions fails, the rest will not run. To identify which assertion failed, a custom error message can be specified as the 3rd param. These error messages can be added later and are not a blocker for merge.

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -27,27 +49,170 @@ public static function wpSetUpBeforeClass() {
+			),
+			'wp_template_part'
+		);
+		$this->assertEquals( get_stylesheet() . '//header', $template_part->id );
+		$this->assertEquals( get_stylesheet(), $template_part->theme );
+		$this->assertEquals( 'header', $template_part->slug );
+		$this->assertEquals( 'publish', $template_part->status );
+		$this->assertEquals( 'theme', $template_part->source );
+		$this->assertEquals( 'header', $template_part->title );
+		$this->assertEquals( '', $template_part->description );
+		$this->assertEquals( 'wp_template_part', $template_part->type );
+		$this->assertEquals( WP_TEMPLATE_PART_AREA_HEADER, $template_part->area );
```
With PHP getting more and more strict, an effort is ongoing to replace all `assertEquals` with `assertSame`.
```suggestion
		$this->assertSame( get_stylesheet() . '//header', $template_part->id );
		$this->assertSame( get_stylesheet(), $template_part->theme );
		$this->assertSame( 'header', $template_part->slug );
		$this->assertSame( 'publish', $template_part->status );
		$this->assertSame( 'theme', $template_part->source );
		$this->assertSame( 'header', $template_part->title );
		$this->assertSame( '', $template_part->description );
		$this->assertSame( 'wp_template_part', $template_part->type );
		$this->assertSame( WP_TEMPLATE_PART_AREA_HEADER, $template_part->area );
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -27,27 +49,170 @@ public static function wpSetUpBeforeClass() {
-		$this->assertSame( 'My Template', $template->title );
-		$this->assertSame( 'Description of my template', $template->description );
-		$this->assertSame( 'wp_template', $template->type );
+		$this->assertEquals( get_stylesheet() . '//my_template', $template->id );
+		$this->assertEquals( get_stylesheet(), $template->theme );
+		$this->assertEquals( 'my_template', $template->slug );
+		$this->assertEquals( 'publish', $template->status );
+		$this->assertEquals( 'custom', $template->source );
+		$this->assertEquals( 'My Template', $template->title );
+		$this->assertEquals( 'Description of my template', $template->description );
+		$this->assertEquals( 'wp_template', $template->type );
```
See note above. I'd suggest leaving these as `assertSame()`.

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -27,27 +49,170 @@ public static function wpSetUpBeforeClass() {
+			'wp_template_part'
+		);
+		$this->assertNotWPError( $template_part );
+		$this->assertEquals( get_stylesheet() . '//my_template_part', $template_part->id );
+		$this->assertEquals( get_stylesheet(), $template_part->theme );
+		$this->assertEquals( 'my_template_part', $template_part->slug );
+		$this->assertEquals( 'publish', $template_part->status );
+		$this->assertEquals( 'custom', $template_part->source );
+		$this->assertEquals( 'My Template Part', $template_part->title );
+		$this->assertEquals( 'Description of my template part', $template_part->description );
+		$this->assertEquals( 'wp_template_part', $template_part->type );
+		$this->assertEquals( WP_TEMPLATE_PART_AREA_HEADER, $template_part->area );
```
Same here:
```suggestion
		$this->assertSame( get_stylesheet() . '//my_template_part', $template_part->id );
		$this->assertSame( get_stylesheet(), $template_part->theme );
		$this->assertSame( 'my_template_part', $template_part->slug );
		$this->assertSame( 'publish', $template_part->status );
		$this->assertSame( 'custom', $template_part->source );
		$this->assertSame( 'My Template Part', $template_part->title );
		$this->assertSame( 'Description of my template part', $template_part->description );
		$this->assertSame( 'wp_template_part', $template_part->type );
		$this->assertSame( WP_TEMPLATE_PART_AREA_HEADER, $template_part->area );
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -27,27 +49,170 @@ public static function wpSetUpBeforeClass() {
+			'<!-- wp:template-part {"slug":"header","align":"full","tagName":"header","className":"site-header","theme":"%s"} /-->',
+			get_stylesheet()
+		);
+		$this->assertEquals( $expected, $template_content );
```
```suggestion
		$this->assertSame( $expected, $template_content );
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -27,27 +49,170 @@ public static function wpSetUpBeforeClass() {
+			'<!-- wp:group --><!-- wp:template-part {"slug":"header","align":"full","tagName":"header","className":"site-header","theme":"%s"} /--><!-- /wp:group -->',
+			get_stylesheet()
+		);
+		$this->assertEquals( $expected, $template_content );
```
```suggestion
		$this->assertSame( $expected, $template_content );
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -27,27 +49,170 @@ public static function wpSetUpBeforeClass() {
+			$content_with_existing_theme_attribute,
+			$theme
+		);
+		$this->assertEquals( $content_with_existing_theme_attribute, $template_content );
```
```suggestion
		$this->assertSame( $content_with_existing_theme_attribute, $template_content );
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -27,27 +49,170 @@ public static function wpSetUpBeforeClass() {
+			$content_with_no_template_part,
+			$theme
+		);
+		$this->assertEquals( $content_with_no_template_part, $template_content );
```
```suggestion
		$this->assertSame( $content_with_no_template_part, $template_content );
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -56,16 +221,27 @@ function test_build_template_result_from_post() {
-		$this->assertSame( 'publish', $template->status );
-		$this->assertSame( 'custom', $template->source );
-		$this->assertSame( 'wp_template', $template->type );
+		$this->assertEquals( $id, $template->id );
+		$this->assertEquals( get_stylesheet(), $template->theme );
+		$this->assertEquals( 'my_template', $template->slug );
+		$this->assertEquals( 'publish', $template->status );
+		$this->assertEquals( 'custom', $template->source );
+		$this->assertEquals( 'wp_template', $template->type );
```
See above. I'd suggest leaving these as `assertSame()`.

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -56,16 +221,27 @@ function test_build_template_result_from_post() {
+		// Test template parts.
+		$id       = get_stylesheet() . '//' . 'my_template_part';
+		$template = get_block_template( $id, 'wp_template_part' );
+		$this->assertEquals( $id, $template->id );
+		$this->assertEquals( get_stylesheet(), $template->theme );
+		$this->assertEquals( 'my_template_part', $template->slug );
+		$this->assertEquals( 'publish', $template->status );
+		$this->assertEquals( 'custom', $template->source );
+		$this->assertEquals( 'wp_template_part', $template->type );
+		$this->assertEquals( WP_TEMPLATE_PART_AREA_HEADER, $template->area );
```
```suggestion
		$this->assertSame( $id, $template->id );
		$this->assertSame( get_stylesheet(), $template->theme );
		$this->assertSame( 'my_template_part', $template->slug );
		$this->assertSame( 'publish', $template->status );
		$this->assertSame( 'custom', $template->source );
		$this->assertSame( 'wp_template_part', $template->type );
		$this->assertSame( WP_TEMPLATE_PART_AREA_HEADER, $template->area );
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -84,14 +260,53 @@ static function( $template ) {
 		$templates    = get_block_templates( array( 'slug__in' => array( 'my_template' ) ), 'wp_template' );
 		$template_ids = get_template_ids( $templates );
-		$this->assertSame( array( get_stylesheet() . '//' . 'my_template' ), $template_ids );
+		$this->assertEquals( array( get_stylesheet() . '//' . 'my_template' ), $template_ids );
```
I'd suggest leaving `assertSame()`.

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -84,14 +260,53 @@ static function( $template ) {
 		$templates    = get_block_templates( array( 'wp_id' => self::$post->ID ), 'wp_template' );
 		$template_ids = get_template_ids( $templates );
-		$this->assertSame( array( get_stylesheet() . '//' . 'my_template' ), $template_ids );
+		$this->assertEquals( array( get_stylesheet() . '//' . 'my_template' ), $template_ids );
```
`assertSame`
```suggestion
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -84,14 +260,53 @@ static function( $template ) {
+		$blocks                             = parse_blocks( $content_template_part_inside_group );
+		$actual                             = _flatten_blocks( $blocks );
+		$expected                           = array( $blocks[0], $blocks[0]['innerBlocks'][0] );
+		$this->assertEquals( $expected, $actual );
```
```suggestion
		$this->assertSame( $expected, $actual );
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -84,14 +260,53 @@ static function( $template ) {
+		$blocks   = parse_blocks( $content_template_part_inside_group_inside_group );
+		$actual   = _flatten_blocks( $blocks );
+		$expected = array( $blocks[0], $blocks[0]['innerBlocks'][0], $blocks[0]['innerBlocks'][0]['innerBlocks'][0] );
+		$this->assertEquals( $expected, $actual );
```
```suggestion
		$this->assertSame( $expected, $actual );
```

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -84,14 +260,53 @@ static function( $template ) {
+		$blocks                       = parse_blocks( $content_without_inner_blocks );
+		$actual                       = _flatten_blocks( $blocks );
+		$expected                     = array( $blocks[0] );
+		$this->assertEquals( $expected, $actual );
```
```suggestion
		$this->assertSame( $expected, $actual );
```

hellofromtonya commented:
My comments/suggestions in the tests are not blockers as I can be update them after the merge.

hellofromtonya commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -488,6 +488,54 @@ public function get_settings() {
+	 */
+	public function get_custom_templates() {
+		$custom_templates = array();
+		if ( ! isset( $this->theme_json['customTemplates'] ) ) {
```
If this element exists, will it always be an array (or iterable)?

If it could be `null`, an empty string, `false` or anything else, the `foreach` will fail and a PHP Warning will be thrown (see in action here https://3v4l.org/rqd0S). May want to add a check here to guard it.

noisysocks replied:
Will defer to @ockham again here 🙂

ockham replied:
Thanks, added in 345ffa8.

hellofromtonya commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -488,6 +488,54 @@ public function get_settings() {
+	 */
+	public function get_template_parts() {
+		$template_parts = array();
+		if ( ! isset( $this->theme_json['templateParts'] ) ) {
```
same note here as above

ockham replied:
345ffa8

hellofromtonya commented:
Whoopsie didn't mean to close this. Reopening.

hellofromtonya commented:
In the tests, there are improvements that can be made which are not blockers to merge this PR:

* With PHP getting more and more strict, an effort is ongoing to replace all `assertEquals` with `assertSame`.
* With multiple assertions in a single test, if one fails, the other assertions do not run and it can be difficult to know which assertion failed. An effort is ongoing to add custom messages to each of these assertions to help with debugging.

I deleted all of the code suggestions for the tests (too much noise). I'll check after merge for any improvements, which BTW can happen after feature freeze.

ockham commented:
FWIW, `get_query_pagination_arrow` was added to GB by this PR: https://github.com/WordPress/gutenberg/pull/33656

ntsekouras commented:
>After rebasing, TT1 Blocks fatals on the frontend since get_query_pagination_arrow is missing. That function is indeed absent from Core; whereas in GB, it’s defined in the WP 5.8 compat layer. I’m a bit confused; I thought that means that it’s been present in Core since 5.8?

That's my bad, sorry 😞 . I will have a PR for adding this in `5.9` for GB. It needs to be added in core as well.

ntsekouras commented:
Added `get_query_pagination_arrow` in `blocks.php`.

ockham commented:
No worries! Thanks a lot for the fix, Nik!

ockham commented:
Thanks to everyone who has worked on this, by adding and improving code, or reviewing!

I think we've addressed all feedback, so I'm setting this PR to ready for review. All that's left is some final approval now 😄

ockham commented:
Ah, rebasing on `trunk` (which now includes https://github.com/WordPress/wordpress-develop/pull/1808) introduced a conflict (duplicated `get_custom_templates()`). I'll fix...

noisysocks commented:
Thanks @bernie! I'll rebase and commit this.

noisysocks commented:
Going to ignore the lint failures as they are failing on `trunk` too.

https://wordpress.slack.com/archives/C02RQBWTW/p1636411775055600
